lib/worker/src/routes/vaults.rs

```diff
@@ -589,3 +589,54 @@ pub struct WriteEventParams {
     timestamp: Option<i64>,
     signature: Option<String>,
 }
+
+async fn upload_stream(
+    uploader: ResumableUploadClient,
+    stream: &mut (impl Stream<Item = Result<impl warp::Buf, warp::Error>> + Unpin + Send + Sync),
+    size: u64,
+) -> basin_common::errors::Result<Hasher> {
+    let mut hasher = Hasher::new();
+
+    const CHUNK_SIZE: usize = 8 * 1024 * 1024; // 8MiB. Must be multiple of 256KiB.
+    let mut received: u64 = 0;
+    let mut collected: Vec<u8> = Vec::new();
+    loop {
+        let first_byte = received;
+        while let Some(buf) = stream.next().await {
+            let mut buf = buf.unwrap();
+            while buf.remaining() > 0 {
+                let chunk = buf.chunk();
+                let chunk_len = chunk.len();
+                collected.extend_from_slice(chunk);
+                buf.advance(chunk_len);
+            }
+
+            if collected.len() > CHUNK_SIZE {
+                break;
+            }
+        }
+
+        let payload = collected
+            .clone()
+            .into_iter()
+            .take(std::cmp::min(CHUNK_SIZE, collected.len()))
+            .collect::<Vec<_>>();
+        received += payload.len() as u64;
+        let last_byte = received - 1;
+        let chunk_size = ChunkSize::new(first_byte, last_byte, Some(size));
+
+        uploader
+            .upload_multiple_chunk(payload.clone(), &chunk_size)
+            .await
+            .map_err(|e| basin_common::errors::Error::Upload(e.to_string()))?;
+
+        hasher.update(&payload);
```

Comment:
one point to note here is that for the verifier to verify the hash and signature it needs to know the payload size otherwise correct hash cannot be calculated.

Reply:
Why's that?

Reply:
But isn't what matters for verification, only the size of the file I want to verify? There's no way to guarantee a fixed buffer size in the current approach, whatever size I pick, there's no guarantee the size of the file is a multiple of the picked value. In this case, I picked 8Mib because it was the GCS recommendation. 

In any case, hashing the file before seems like a good idea, so we always do one operation to the Hasher.

Reply:
yes, makes sense. the verifiers should care about the hashes of individual chunks of file. also, the assumption about he buffer size is not the correct assumption. make sense to use the old scheme for hashing the file